scipy/optimize/_sda.py

```diff
@@ -0,0 +1,568 @@
+        self.temperature_restart = 0.1
+        # VisitingDistribution instance
+        vd = VisitingDistribution(lower, upper, qv, self.rs)
+        # Markov chain instance
+        self.mc = MarkovChain(qa, vd, self.owf, self.rs, self.es)
+        self.qv = qv
+        self.pure_sa = pure_sa
+
+    def search(self):
+        max_steps_reached = False
+        self._iter = 0
+        t1 = np.exp((self.qv - 1) * np.log(2.0)) - 1.0
+        while(not max_steps_reached):
+            for i in range(self.max_steps):
+                # Compute temperature for this step
+                s = float(i) + 2.0
+                t2 = np.exp((self.qv - 1) * np.log(s)) - 1.0
+                temperature = self.temperature_start * t1 / t2
+                self._iter += 1
+                if self._iter == self.max_steps:
+                    max_steps_reached = True
+                    break
+                # Need a re-annealing process?
+                if temperature < self.temperature_restart:
+                    self.es.reset(self.owf, self.rs)
+                    break
+                # starting Markov chain
+                self.mc.run(i, temperature)
+                if self.owf.nb_fun_call >= self.maxfun:
+                    break
+                if not self.pure_sa:
+                    self.mc.local_search()
+                    if self.owf.nb_fun_call >= self.maxfun:
+                        break
+
+    @property
+    def result(self):
+        """ The OptimizeResult """
+        res = OptimizeResult()
+        res.x = self.es.xbest
+        res.fun = self.es.ebest
+        res.nit = self._iter
+        res.ncall = self.owf.nb_fun_call
+        return res
+
+
+def sda(func, x0, bounds, maxiter=1000, minimizer_kwargs=None,
+        initial_temp=5230., visit=2.62, accept=-5.0, maxfun=1e7, seed=None,
+        pure_sa=False):
+    """
+    Find the global minimum of a function using the Simulated Dual Annealing
+    algorithm
+
+    Parameters
+    ----------
+    func : callable
+        The objective function to be minimized.  Must be in the form
+        ``f(x, *args)``, where ``x`` is the argument in the form of a 1-D array
+        and ``args`` is a  tuple of any additional fixed parameters needed to
+        completely specify the function.
+    x0 : ndarray
+        The starting coordinates. If ``None`` is provided, initial
+        coordinates are automatically generated.
+    bounds : sequence
+        Bounds for variables.  ``(min, max)`` pairs for each element in ``x``,
+        defining the lower and upper bounds for the optimizing argument of
+        `func`. It is required to have ``len(bounds) == len(x)``.
+        ``len(bounds)`` is used to determine the number of parameters in ``x``.
+    maxiter : int, optional
+        The maximum number of sda iterations. Increase this value if the
+        objective function is very complicated with high dimensions.
+    minimizer_kwargs : dict, optional
```

Comment:
I think you should change the name of this argument to something like "local_search_options". It sounds much better and gives a better notion of what this argument is for.

Reply:
Thanks for this. I was tempted to use the same argument name that basin hopping is offering but if you think this is more easy to read, I change it to "local_search_options"